### PR TITLE
Fcc 60 sub themes

### PR DESCRIPTION
**Pull request by jamesseth**
Add CRUD functionality for Theme and Sub-themes

+ Resources
    + add_theme.py.............endpoint to add a new theme
    + get_themes.py............endpoint to fetch themes
    + delete_theme.py..........endpoint to delete a theme
    + rename_theme.py..........endpoint to rename a theme

    + add_theme.py.............endpoint to add a new sub theme
    + get_sub_themes.py........endpoint to fetch sub themes
    + delete_subtheme.py.......endpoint to delete a sub theme
    + rename_subtheme.py.......endpoint to rename a sub theme

 +Modified
    + models/theme.py..........Type hinting and docstrings added
                               DB session flush removed on delete
    + app.py...................Resources imported and endpoints added

### CODE REVIEW

File: Analytics/resources/themes/get_themes.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        """
+        # Get arguments passed in POST request
+        args = self.reqparser.parse_args()
+
+        # pdb.set_trace()
+        # fetch themes using arguments passed in post request
+        themes = []
+        if "id" in args:
+            theme = Theme.get_by_id(args["id"])
+            if theme:
```
Comment (DannyLee12) [inline Analytics/resources/themes/get_themes.py]: Consider refactoring this. Don't need two copies of
```python
            if theme:
                themes.append(theme.json())
```
Rather do:

```python
if "name" in args:
    theme = Theme.get_by_name(args["name"])
elif "id" in args:
    theme = Theme.get_by_id(args["id"])

theme.append(theme.json())
```
Or something like that.

File: Analytics/resources/themes/rename_subtheme.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        self.reqparser = reqparse.RequestParser()
+        self.reqparser.add_argument('current_name', required=True, type=str, help='Current theme name required',
+                                    location=['form', 'json'])
+        self.reqparser.add_argument('new_name', required=True, type=str, help='New theme name required',
+                                    location=['form', 'json'])
+
+    @jwt_required
+    def post(self) -> ({str: str}, HTTPStatus):
+        """
+        renames a existing sub theme entry in the database
```
Comment (DannyLee12) [inline Analytics/resources/themes/rename_subtheme.py]: Rename

File: Analytics/resources/themes/rename_subtheme.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        :post_argument  current_name: the name of the sub theme to rename
```
Comment (DannyLee12) [inline Analytics/resources/themes/rename_subtheme.py]: typehints are usually:

```python
:param post_argument: definintion

File: Analytics/resources/themes/rename_theme.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+from flask_jwt_extended import jwt_required, get_jwt_claims
+from flask_restful import Resource
+from flask_restful import reqparse
+
+from models.theme import Theme
+
+
+class RenameTheme(Resource):
+    """
+    renames a existing Theme entry in the database
```
Comment (DannyLee12) [inline Analytics/resources/themes/rename_theme.py]: Rename

File: Analytics/resources/themes/rename_theme.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    def __init__(self) -> None:
+        """
+        Sets the required arguments to be in the POST request
```
Comment (DannyLee12) [inline Analytics/resources/themes/rename_theme.py]: Set

File: Analytics/resources/themes/test_subthemes.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            self.theme = Theme("_test_add_Subtheme_")
+            self.theme.save()
+            self.theme.commit()
+            self.theme = Theme.get_by_name("_test_add_Subtheme_")
+        self.subtheme = self.create_dummy_subtheme()
+
+
+    def create_test_client(self) -> (FlaskClient, AppContext):
+        """
+        Creates a flask client for testing
```
Comment (DannyLee12) [inline Analytics/resources/themes/test_subthemes.py]: Create

File: Analytics/resources/themes/add_subtheme.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    def __init__(self) -> None:
+        """
+        Set required arguments for POST request
+        """
+        self.reqparser = reqparse.RequestParser()
+        self.reqparser.add_argument('theme', required=False, store_missing=False, default=None, type=str,
+                                    help='Theme required', location=['form', 'json'])
+        self.reqparser.add_argument('theme_id', required=False, store_missing=False, default=None, type=str,
+                                    help='Theme required', location=['form', 'json'])
```
Comment (j0rdanl0urens) [inline Analytics/resources/themes/add_subtheme.py]: Is it right theme and theme id have the same hint?

File: Analytics/resources/themes/add_subtheme.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    """
+
+    def __init__(self) -> None:
+        """
+        Set required arguments for POST request
+        """
+        self.reqparser = reqparse.RequestParser()
+        self.reqparser.add_argument('theme', required=False, store_missing=False, default=None, type=str,
+                                    help='Theme required', location=['form', 'json'])
+        self.reqparser.add_argument('theme_id', required=False, store_missing=False, default=None, type=str,
```
Comment (j0rdanl0urens) [inline Analytics/resources/themes/add_subtheme.py]: It says in post that this should be int not str

File: Analytics/resources/themes/test_subthemes.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        self.auth_header = self.get_auth_header()
+
+        self.theme = Theme.get_by_name("_test_add_Subtheme_")
+        if not self.theme:
+            self.theme = Theme("_test_add_Subtheme_")
+            self.theme.save()
+            self.theme.commit()
+            self.theme = Theme.get_by_name("_test_add_Subtheme_")
+        self.subtheme = self.create_dummy_subtheme()
+
```
Comment (j0rdanl0urens) [inline Analytics/resources/themes/test_subthemes.py]: 1 line

File: Analytics/resources/themes/test_subthemes.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+    def create_test_client(self) -> (FlaskClient, AppContext):
+        """
+        Create flask testing client
```
Comment (j0rdanl0urens) [inline Analytics/resources/themes/test_subthemes.py]: Add missing return type

File: Analytics/resources/themes/test_subthemes.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        response = self.client.post('/admin/themes/rename_subtheme', json={ "current_name": self.subtheme.name,
+                                                                            "new_name": "_____________________"
+                                                                            }, headers=self.auth_header)
+        self.assertEqual(response.status_code, HTTPStatus.OK)
+        response = response.get_json()
+        self.assertEqual(response["id"], self.subtheme.id)
+        self.assertEqual(response["message"], "Subtheme renamed")
+        self.assertEqual(response["old_name"], "_TEST_SUB_THEME_")
+        self.assertEqual(response["new_name"], "_____________________")
+
```
Comment (j0rdanl0urens) [inline Analytics/resources/themes/test_subthemes.py]: 1 line

File: Analytics/resources/themes/test_subthemes.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+        Delete a SubTheme by id and check the client response status code for http status 204 (NO_CONTENT)
+        """
+        if not self.subtheme:
+            self.subtheme = self.create_dummy_subtheme()
+        response = self.client.post('/admin/themes/delete_subtheme', json={"id": self.subtheme.id}, headers=self.auth_header)
+        self.assertEqual(response.status_code, HTTPStatus.NO_CONTENT)
+
+    def tearDown(self):
+        """ Handle the cleanup after tests"""
+
```
Comment (j0rdanl0urens) [inline Analytics/resources/themes/test_subthemes.py]: 1 line

File: Analytics/resources/themes/add_theme.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+                  A JSON with an appropriate error message and http status applicable to the error
+        """
+        if not get_jwt_claims()['admin']:
+            return {"error": "administration privileges required"}, HTTPStatus.FORBIDDEN
+
+        # Get arguments
+        args = self.reqparser.parse_args()
+
+        # Check the theme name is not empty, abort if it is empty
+        if args["name"] == "":
```
Comment (j0rdanl0urens) [inline Analytics/resources/themes/add_theme.py]: Rather do `if not args["name"]:`
You do this in a lot of places, please change them all.

**Pull request merged**